schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,298 @@
+
+By default, each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links. The schema and algorithm presented here also allows for inline child nodes rather than links, with read operations able to traverse multiple nodes within a single block where they are inlined. The production of inlined IPLD HashMaps is left unspecified and users should be aware that inlining breaks canonical form guarantees.
+
+## Structure
+
+### Parameters
+
+Configurable parameters for any given IPLD HashMap:
+
+* `hashAlg`: The hash algorithm applied to keys in order to evenly distribute entries throughout the data structure. The algorithm is chosen based on speed, `digestLength` and randomness properties (but it must be available to the reader, hence the need for shared defaults, see below).
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An Array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield, stored as Bytes, where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
```

Comment:
Shouldn't this be a fixed size?

Reply:
@dignifiedquire: it's dependent on `bitWidth` but @warpfork has been very firm on his desire to _not_ introduce dependent types into IPLD schemas so we can't represent that here. 

I've added this to the notes below the schema:

> * The size of `map` is determined by `bitWidth` since it holds one bit per possible data element. It must be `1` or `2`<sup>`bitWidth`</sup>` / 8` bytes long, whichever is largest.

Reply:
I am confused, is this a different indexing scheme than used in go-hamt-ipld? 
Because looking at https://github.com/ipfs/go-hamt-ipld/blob/master/hamt.go#L73-L74 it needs to always be of size 256 bits.

Reply:
Yes, this spec is not intended to reflect go-hamt-ipld. It has a flexible `bitWidth`, the go-hamt-ipld data structure is fixed to 8-bits for simplicity. Fixing to `8` gives a fixed `map` size of `32` bytes and a maximum number of data elements of `256`. The reference implementation for the purposes of writing this spec is https://github.com/rvagg/iamap (it's currently missing the ability to deal with inlined child nodes).